src/boot/efi/meson.build

```diff
@@ -258,26 +250,52 @@ foreach arg : get_option('c_args')
         endif
 endforeach
 
-efi_ldflags = ['-T', efi_lds,
-               '-shared',
-               '-Bsymbolic',
-               '-nostdlib',
-               '--no-undefined',
-               '--warn-common',
-               '--fatal-warnings',
-               '-znocombreloc',
-               '--build-id=sha1',
-               '-L', efi_libdir,
-               efi_crt0]
+efi_ldflags = [
+        '-fuse-ld=' + get_option('efi-ld'),
+        '-L', efi_libdir,
+        '-nostdlib',
+        '-shared',
+        '-T', efi_lds,
+        '-Wl,--build-id=sha1',
+        '-Wl,--fatal-warnings',
+        '-Wl,--no-undefined',
+        '-Wl,--warn-common',
+        '-Wl,-Bsymbolic',
+        '-z', 'nocombreloc',
+        efi_crt0,
+]
 if efi_arch[1] in ['aarch64', 'arm', 'riscv64']
         # Aarch64, ARM32 and 64bit RISC-V don't have an EFI capable objcopy.
         # Use 'binary' instead, and add required symbols manually.
-        efi_ldflags += ['--defsym=EFI_SUBSYSTEM=0xa']
+        efi_ldflags += ['-Wl,--defsym=EFI_SUBSYSTEM=0xa']
         efi_format = ['-O', 'binary']
 else
         efi_format = ['--target=efi-app-@0@'.format(efi_arch[1])]
 endif
 
+if run_command('grep', '-q', '__CTOR_LIST__', efi_lds).returncode() == 0
+        # fedora has a patched gnu-efi that adds support for ELF constructors.
+        # If ld is called by gcc something about these symbols breaks, resulting
+        # in sd-boot freezing when gnu-efi runs the constructors. Force defining
+        # them seems to work around this.
+        efi_ldflags += [
+                '-Wl,--defsym=_init_array=0',
+                '-Wl,--defsym=_init_array_end=0',
+                '-Wl,--defsym=_fini_array=0',
+                '-Wl,--defsym=_fini_array_end=0',
+                '-Wl,--defsym=__CTOR_LIST__=0',
+                '-Wl,--defsym=__CTOR_END__=0',
+                '-Wl,--defsym=__DTOR_LIST__=0',
+                '-Wl,--defsym=__DTOR_END__=0',
```

Comment:
constructors are a terrible thing. I don#t want to know what fedora wants with them...


what happens if we define these symbols unconditionally?

Reply:
That would just mean having some superfluous symbols. Probably. I don't think it even matters what they are since the fedora gnu-efi code only takes their address: https://src.fedoraproject.org/rpms/gnu-efi/blob/rawhide/f/0027-Make-ELF-constructors-and-destructors-work.patch#_602